source/constructs/lib/agent/unstructured-crawler/UnstructuredCrawler.py

```diff
@@ -0,0 +1,156 @@
+    
+    Args:
+        s3_client: boto3 s3 client
+        bucket_name: name of the s3 bucket
+        scan_depth: number of files to be scanned
+        include_file_extensions: list of file extensions to be included in the scan
+        exclude_file_extensions: list of file extensions to be excluded in the scan
+        prefix: prefix of the s3 bucket
+    """
+    bucket_info = {}
+    detection_files, include_files, exclude_files = {}, {}, {}
+    paginator = s3_client.get_paginator('list_objects_v2')
+    pages = paginator.paginate(Bucket=bucket_name, Prefix=prefix, PaginationConfig={'PageSize': 1000})
+
+    supported_types = []
+    for file_extension in supported_text_types + suported_image_types:
+        supported_types.append(file_extension)
+
+    # iterate over pages
+    for page in pages:
+        # loop through objects in page
+        if 'Contents' in page:
+            page_detection_files, page_include_files, page_exclude_files = summarize_page(page, supported_types, include_file_extensions, exclude_file_extensions)
+            detection_files = update_dict_files(detection_files, page_detection_files)
+            include_files = update_dict_files(include_files, page_include_files)
+            exclude_files = update_dict_files(exclude_files, page_exclude_files)
+
+    detection_files = postprocess_crawler_result(detection_files, scan_depth)
+    include_files = postprocess_crawler_result(include_files, scan_depth)
+    exclude_files = postprocess_crawler_result(exclude_files, scan_depth)
+
+    # Get the current UTC time
+    now_utc = datetime.datetime.now(datetime.timezone.utc)
+    utc_string = now_utc.strftime('%Y-%m-%d %H:%M:%S.%f %Z')
+
+    bucket_info['bucket_name'] = bucket_name
+    bucket_info['last_updated_time'] = utc_string
+    bucket_info['detection_files'] = detection_files
+    bucket_info['include_files'] = include_files
+    bucket_info['exclude_files'] = exclude_files
+    bucket_info['include_file_extensions'] = include_file_extensions
+    bucket_info['exclude_file_extensions'] = exclude_file_extensions
+
+    return bucket_info
+
+def upload_bucket_info(s3_client, bucket_info, source_bucket_name, result_bucket_name, prefix=''):
+    # dump json format content to a file and save to s3
+    json_file_path = '/tmp/json_file.json'
+    with open(json_file_path, 'w') as json_file:
+        json.dump(bucket_info, json_file, ensure_ascii=False)
+    s3_client.upload_file(json_file_path, result_bucket_name, f"{prefix}/{source_bucket_name}_info.json")
+
+    # delete the tmp file
+    os.remove(json_file_path)
+
+
+def lambda_handler(event, context):
+    # print(event)
+    source_bucket_name = event['SourceBucketName']
+    result_bucket_name = event['ResultBucketName']
+    scan_depth = int(event['ScanDepth'])
+    include_file_extensions = event['IncludeFileExtensions']
+    exclude_file_extensions = event['ExcludeFileExtensions']
+    s3 = boto3.client('s3')
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

A Boto3 client is initialized from a Lambda handler. To speed up your Boto3 client initialization and minimize the operational cost of your Lambda function, we recommend that you create the client at the level of the module that contains the handler, and then reuse it between invocations.

[Learn more](https://docs.aws.amazon.com/lambda/latest/dg/best-practices.html#function-code)

Reply:
fixed